contrib/epee/include/misc_language.h

```diff
@@ -96,7 +96,7 @@ namespace misc_utils
   template<class t_scope_leave_handler>
   auto_scope_leave_caller create_scope_leave_handler(t_scope_leave_handler f)
   {
-    auto_scope_leave_caller slc(new call_befor_die<t_scope_leave_handler>(f));
+    auto_scope_leave_caller slc = std::make_shared<call_befor_die<t_scope_leave_handler>>(f);
```

Comment:
While we're making this change - the variable `slc` can be removed and `f` should be moved (although I don't think `call_befre_die` properly moves either).

Reply:
I realize the compiler will (mostly likely) do the same thing for the named variable, but why even have it?

And your callback example is too trivial to make a difference - try with a non-trivial destructor in the callback. I doubt the compiler can "assume" its a move unless specifically told to do so.

Reply:
The compiler is able to apply return-value-optimization even in that case too [1, 2]. 

As you can see from the diff page in [1, 2], both versions, even for complex types are identical. (Take a look at RVO in [3]).

1. https://godbolt.org/z/7Pr3f4xdf  
2. https://godbolt.org/z/1hTKMd9Kc
3. https://sigcpp.github.io/2020/06/08/return-value-optimization

Reply:
Now that being said, I don't have any issue with changing the code to :
```
  return std::make_shared<call_befor_die<t_scope_leave_handler>>(f);
```

My point is it is matter of cosmetic. So if you think that is better, let me know, I will change it to that.

Reply:
I know the compiler will almost certainly apply NVRO, which is why I said: "will (mostly likely) do the same thing". That was in reference to the named value in the return. It is purely cosmetic, and I would've just dropped it since it useless. Keep it for historical reasons if you want.

The other comment was about `std::move`. I should've started a separate issue for that, to avoid confusion. Your example isn't highlighting the problem easily, because copy and moves are similar in your example. Remove the definition (leaving just the declaration), and you can easily see that the copy constructor is being called twice for this function, when it could be calling the move constructor twice. In some cases the two are identical, but not in all cases.

Reply:
>The other comment was about std::move. I should've started a separate issue for that, to avoid confusion. Your example isn't highlighting the problem easily, because copy and moves are similar in your example. Remove the definition (leaving just the declaration), and you can easily see that the copy constructor is being called twice for this function, when it could be calling the move constructor twice. In some cases the two are identical, but not in all cases.

I see what you saying, but that is totally different issue than what I am trying to solve in this PR. My main goal is to **remove the extra memory allocation happening** here. There are a lot of areas that `std::move` can help with optimization. But that does not fall into this PR area.


By the way, no need to remove the definition. Adding simple `__attribute__ ((noinline))` would show the copy constructor gets called in both ways. So we can say not directly related to this PR anyway.

1. https://godbolt.org/z/P9nrbzoPc